static_frame/core/index_hierarchy.py

```diff
@@ -1051,14 +1050,65 @@ def gen() -> tp.Iterator[tp.Tuple[tp.Hashable, ...]]:
                 )
 
 
-    # def relabel_at_depth(self: IH,
-    #         mapper: RelabelInput,
-    #         depth_level: DepthLevelSpecifier = 0
-    #         ) -> IH:
-    #     '''
-    #     Return a new :obj:`IndexHierarchy` after applying the hte `mapper` to the depth level or levels specified by `depth_level`.
-    #     '''
-    #     pass
+    def relabel_at_depth(self,
+            mapper: RelabelInput,
+            depth_level: DepthLevelSpecifier = 0
+            ) -> "IndexHierarchy":
+        '''
+        Return a new :obj:`IndexHierarchy` after applying the the `mapper` to the depth level or levels specified by `depth_level`.
+        '''
+        if self._recache:
+            self._update_array_cache()
+
+        if isinstance(depth_level, INT_TYPES):
+            # Single depth levels have an optimized path through IndexLevels
+            return self.__class__(
+                    levels=self._levels.relabel_at_specific_depth(mapper, depth_level),
+                    name=self._name,
+                    )
+
+        depth_level = list(depth_level)
+
+        if len(depth_level) == 1:
+            # Optimized path for single depth level
+            return self.__class__(
+                levels=self._levels.relabel_at_specific_depth(mapper, depth_level[0]),
+                name=self._name,
+                )
+        elif len(depth_level) == self.depth:
+            # Equivalent to relabel!
+            return self.relabel(mapper)
+
+        index_constructors = tuple(self._levels.index_types())
+
+        def get_modified(array_items: tuple) -> tp.Iterable[tp.Any]: # type: ignore
+            if callable(mapper):
+                return mapper(array_items) # type: ignore
+
+            # if a mapper, it must support both __getitem__ and __contains__
+            return mapper.get(array_items, array_items) # type: ignore
+
+        array_getter = itemgetter(*depth_level)
+
+        def gen() -> tp.Iterator[np.ndarray]:
+            for array in self._blocks.axis_values(axis=1):
```

Comment:
This will force type consolidation across the columns of the IH, which can lead to some undesirable coercions. If we make this function only operate on elements of each depth level in isolation, we can avoid this and perhaps use / extend your approach on `self._levels`.

Reply:
This is the exact same approach used in `IndexHierarchy.relabel`. I opted for this as I could not think of an alternative way to apply a mapping that's row-based on a subset of the index depths.